src/map/battleground.cpp

```diff
@@ -222,59 +621,863 @@ int bg_team_get_id(struct block_list *bl)
+		clif_displaymessage(sd->fd, buf);
+		return BG_APPLY_NONE;
+	}
+
+	if (bg_player_is_in_bg_map(sd)) {
+		clif_displaymessage(sd->fd, msg_txt(sd, 337)); // You may not join a battleground queue when you're in a battleground map.
+		return BG_APPLY_NONE;
+	}
+
+	std::weak_ptr<s_battleground_type> weakptr = bg_search_name(name);
+	std::shared_ptr<s_battleground_type> bg;
+
+	if (weakptr.expired()) {
+		ShowWarning("bg_queue_join_multi: Could not find battleground \"%s\" requested by %s (AID: %d / CID: %d)\n", name, sd->status.name, sd->status.account_id, sd->status.char_id);
+		return BG_APPLY_INVALID_NAME;
+	} else
+		bg = weakptr.lock();
+
+	if (bg->min_lvl && sd->status.base_level < bg->min_lvl)
+		return BG_APPLY_PLAYER_LEVEL; // Level too low
+	
+	if (bg->max_lvl && sd->status.base_level > bg->max_lvl)
+		return BG_APPLY_PLAYER_LEVEL; // Level too high
+
+	if (bg_queues.empty()) {
+		std::shared_ptr<s_battleground_queue> queue = bg_queue_create(bg);
+		bool r = rnd() % 2 != 0;
+
+		if (!r) {
+			while (!list.empty() && queue->teama_members.size() < bg->required_players) {
+				struct map_session_data *sd2 = list.back();
+
+				list.pop_back();
+
+				if (!sd2 || sd2->bg_queue)
+					continue;
+
+				if (!bg_queue_check_joinable(bg, sd2, name))
+					continue;
+
+				sd2->bg_queue = queue;
+				clif_bg_queue_apply_result(BG_APPLY_ACCEPT, name, sd2);
+				clif_bg_queue_apply_notify(name, sd2);
+				queue->teama_members.insert(queue->teama_members.begin(), sd2);
+			}
+		} else {
+			while (!list.empty() && queue->teamb_members.size() < bg->required_players) {
+				struct map_session_data *sd2 = list.back();
+
+				list.pop_back();
+
+				if (!sd2 || sd2->bg_queue)
+					continue;
+
+				if (!bg_queue_check_joinable(bg, sd2, name))
+					continue;
+
+				sd2->bg_queue = queue;
+				clif_bg_queue_apply_result(BG_APPLY_ACCEPT, name, sd2);
+				clif_bg_queue_apply_notify(name, sd2);
+				queue->teamb_members.insert(queue->teamb_members.begin(), sd2);
+			}
+		}
+
+		bg_queues.insert(bg_queues.begin(), queue);
+
+		return BG_APPLY_ACCEPT;
+	} else {
+		std::shared_ptr<s_battleground_queue> queue;
+		bool r = rnd() % 2 != 0;
+
+		for (const auto &it : bg_queues) {
+			try {
+				queue = it;
+			} catch (std::out_of_range &) {
+				continue;
+			}
+
+			if (queue->in_ready_state)
+				continue;
+
+			if (queue->teama_members.size() + list.size() <= bg->required_players || queue->teamb_members.size() + list.size() <= bg->required_players) { // Make sure there's enough space on one side to join as a party/guild in this queue
+				if (!r && queue->teama_members.size() + list.size() <= bg->required_players) {
+					while (!list.empty() && queue->teama_members.size() < bg->required_players) {
+						struct map_session_data *sd2 = list.back();
```

Comment:
Again, what's the reasoning behind not using list.pop_back() here instead of accessing through list.back()? Ditto elsewhere